source/opt/flatten_array_matrix_stage_var.cpp

```diff
@@ -0,0 +1,908 @@
+  }
+
+  // Handle matrix case.
+  if (stage_var_type->opcode() == SpvOpTypeMatrix) {
+    return CreateFlattenedStageVarsForMatrix(stage_var_type, storage_class,
+                                             extra_array_length);
+  }
+
+  // Handle scalar or vector case.
+  NestedCompositeComponents flattened_var;
+  uint32_t type_id = stage_var_type->result_id();
+  if (extra_array_length != 0) {
+    type_id = GetArrayType(type_id, extra_array_length);
+  }
+  uint32_t ptr_type_id =
+      context()->get_type_mgr()->FindPointerToType(type_id, storage_class);
+  uint32_t id = TakeNextId();
+  std::unique_ptr<Instruction> variable(
+      new Instruction(context(), SpvOpVariable, ptr_type_id, id,
+                      std::initializer_list<Operand>{
+                          {SPV_OPERAND_TYPE_STORAGE_CLASS,
+                           {static_cast<uint32_t>(storage_class)}}}));
+  flattened_var.SetSingleComponentVariable(variable.get());
+  context()->AddGlobalValue(std::move(variable));
+  return flattened_var;
+}
+
+Instruction* FlattenArrayMatrixStageVariable::GetTypeOfStageVariable(
+    Instruction* stage_var, bool has_extra_arrayness) {
+  uint32_t pointee_type_id = GetPointeeTypeIdOfVar(stage_var);
+
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  Instruction* type_inst = def_use_mgr->GetDef(pointee_type_id);
+  if (has_extra_arrayness) {
+    assert(type_inst->opcode() == SpvOpTypeArray &&
+           "Stage variable with an extra arrayness must be an array");
+    // Get the type without extra arrayness.
+    uint32_t elem_type_id =
+        type_inst->GetSingleWordInOperand(kOpTypeArrayElemTypeInOperandIndex);
+    type_inst = def_use_mgr->GetDef(elem_type_id);
+  }
+
+  return type_inst;
+}
+
+Instruction* FlattenArrayMatrixStageVariable::GetStageVariable(
+    uint32_t stage_var_id) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  Instruction* stage_var = def_use_mgr->GetDef(stage_var_id);
+  if (stage_var == nullptr) {
+    std::string message("Stage variable does not exist");
+    context()->consumer()(SPV_MSG_ERROR, "", {0, 0, 0}, message.c_str());
+    return nullptr;
+  }
+  if (stage_var->opcode() != SpvOpVariable) {
+    std::string message("Stage variable must be OpVariable instruction");
+    message += "\n  " +
+               stage_var->PrettyPrint(SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES);
+    context()->consumer()(SPV_MSG_ERROR, "", {0, 0, 0}, message.c_str());
+    return nullptr;
+  }
+  return stage_var;
+}
+
+Pass::Status FlattenArrayMatrixStageVariable::Process() {
+  std::unordered_map<uint32_t, StageVariableInfo>
+      stage_var_ids_to_stage_var_info;
+  CollectStageVariablesToFlatten(&stage_var_ids_to_stage_var_info);
+
+  Pass::Status status = Status::SuccessWithoutChange;
+  for (auto itr : stage_var_ids_to_stage_var_info) {
+    Instruction* stage_var = GetStageVariable(itr.first);
+    if (stage_var == nullptr) return Pass::Status::Failure;
```

Comment:
Again assert if this cannot actually happen.  This is only called for pairs that are in `stage_var_ids_to_stage_var_info`.  Something is only inserted there if it is an id that has a location decoration applied to it.  In a valid spir-v module, that has to be an OpVariable.